src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -150,24 +141,64 @@ object ReadSegment {
     }
   }
 
-
   /** Creates a new read segment of the same type but with a new length */
   def apply(segment: ReadSegment, length: Int): ReadSegment = segment match {
     case seg: Template         => Template(seg.offset, length)
     case seg: SampleBarcode    => SampleBarcode(seg.offset, length)
     case seg: MolecularBarcode => MolecularBarcode(seg.offset, length)
     case seg: Skip             => Skip(seg.offset, length)
   }
+
+  def structureRead(segments: Seq[ReadSegment], bases: String, useOffset: Boolean = false): String = {
+    var idx = 0
+    val barcodeBases = new StringBuilder
+    while (idx < segments.length) {
+      val seg = segments(idx)
+      barcodeBases.append(if (useOffset) seg.structureRead(bases, offset = -1) else seg.structureRead(bases, offset = barcodeBases.length))
+      idx += 1
+    }
+    barcodeBases.toString()
+  }
 }
 
 /** Base for all types of segments of a read */
 sealed trait ReadSegment {
-  /** The number of bases in the read preceeding this segment. */
+  /** The number of bases in the read preceding this segment. */
   def offset: Int
   /** The length of this segment. */
   def length: Int
   /** The single-character symbol representing this segment. */
   def symbol: Char
+
+  /** Gets first and last bases (exclusive, 0-based) associated with this read segment.  If strict is false then only
+    * bound the last base to `basesLen`, otherwise if the end is too big, throw an exception.
+    **/
+  private[util] def getStartAndEnd(basesLen: Int, offset: Int = -1, strict: Boolean): (Int, Int) = {
+    val start = if (offset < 0) this.offset else offset
+    val end = length + start
+    if (basesLen < end) {
+      if (strict) throw new IllegalStateException("Not enough bases for read segment: " + this)
+      else (start, basesLen)
+    }
+    else (start, end)
+  }
+
+  /** Gets the bases associated with this read segment.  If strict is false then only return
+    * the sub-sequence for which we have bases in `bases`, otherwise throw an exception.
+    **/
+  def structureRead(bases: String, offset: Int = -1, strict: Boolean = true): String = {
```

Comment:
I know this mirrors the name in `ReadStructure` but I don't think that's appropriate.  Can we call these `extractBases` and `extractBasesAndQualities` or something similar?  I think that's a more accurate description of what they do.